sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryMetrics.java

```diff
@@ -0,0 +1,907 @@
+    private static final ConcurrentHashMap<MeterRegistry, AtomicLong> registryRefCount = new ConcurrentHashMap<>();
+
+    private static String convertStackTraceToString(Throwable throwable)
+    {
+        try (StringWriter sw = new StringWriter();
+             PrintWriter pw = new PrintWriter(sw))
+        {
+            throwable.printStackTrace(pw);
+            return sw.toString();
+        }
+        catch (IOException ioe)
+        {
+            throw new IllegalStateException(ioe);
+        }
+    }
+
+    private static CompositeMeterRegistry createFreshRegistry() {
+        CompositeMeterRegistry registry = new CompositeMeterRegistry();
+        if (logger.isTraceEnabled()) {
+            registry.config().onMeterAdded(
+                (meter) -> {
+                    logger.trace(
+                        "Meter '{}' added. Callstack: {}",
+                        meter.getId().getName(),
+                        convertStackTraceToString(new IllegalStateException("Dummy")));
+                }
+            );
+        }
+
+        return registry;
+    }
+
+    public static void recordSystemUsage(
+        float averageSystemCpuUsage,
+        float freeMemoryAvailableInMB
+    ) {
+        if (compositeRegistry.getRegistries().isEmpty()) {
+            return;
+        }
+
+        DistributionSummary averageSystemCpuUsageMeter = DistributionSummary
+            .builder(nameOf("system.avgCpuLoad"))
+            .baseUnit("%")
+            .description("Avg. System CPU load")
+            .maximumExpectedValue(100d)
+            .publishPercentiles(0.95, 0.99)
+            .publishPercentileHistogram(true)
+            .register(compositeRegistry);
+        averageSystemCpuUsageMeter.record(averageSystemCpuUsage);
+
+        DistributionSummary freeMemoryAvailableInMBMeter = DistributionSummary
+            .builder(nameOf("system.freeMemoryAvailable"))
+            .baseUnit("MB")
+            .description("Free memory available")
+            .publishPercentiles()
+            .publishPercentileHistogram(false)
+            .register(compositeRegistry);
+        freeMemoryAvailableInMBMeter.record(freeMemoryAvailableInMB);
+    }
+
+    public static void recordOperation(
+        CosmosAsyncClient cosmosAsyncClient,
+        CosmosDiagnostics cosmosDiagnostics,
+        int statusCode,
+        Integer maxItemCount,
+        Integer actualItemCount,
+        String containerId,
+        String databaseId,
+        OperationType operationType,
+        ResourceType resourceType,
+        ConsistencyLevel consistencyLevel,
+        String operationId,
+        float requestCharge,
+        Duration latency
+    ) {
+        if (compositeRegistry.getRegistries().isEmpty() ||
+            !clientAccessor.isClientTelemetryMetricsEnabled(cosmosAsyncClient)) {
+            return;
+        }
+
+        boolean isPointOperation = maxItemCount == null || maxItemCount < 0;
```

Comment:
when maxItemCount will be < 0 for pointOperation?

Reply:
No concrete scenario - butsince it isn't nullable just defense in depth for someone using -1 as "null"